Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -4160,6 +4153,32 @@ RefPtr<Frame> FrameLoader::findFrameForNavigation(const AtomString& name, Docume
     return frame;
 }
 
+// https://fetch.spec.whatwg.org/#fetch-scheme
+static bool urlIsFetchScheme(const URL& url)
+{
+    return (url.protocolIsInHTTPFamily() || url.protocolIsAbout() || url.protocolIsBlob() || url.protocolIsData() || url.protocolIsFile());
+}
+
+bool FrameLoader::dispatchNonTraversalNavigateEvent(const URL& newURL, const NavigationAction& action, NavigationHistoryBehavior historyHandling, bool isSameDocument)
+{
+    RefPtr document = m_frame->document();
+    if (!document || !document->settings().navigationAPIEnabled())
+        return true;
+    RefPtr window = document->protectedWindow();
+    if (!window)
+        return true;
+    if (!action.downloadAttribute().isNull()) // Download events are handled later.
+        return true;
+    if (!isSameDocument && !urlIsFetchScheme(newURL))
+        return true;
+
+    auto navigationType = determineNavigationType(m_loadType, historyHandling);
+    if (navigationType == NavigationNavigationType::Traverse) // Handled elsewhere.
```

Comment:
In a function named FrameLoader::dispatchNonTraversalNavigateEvent I'd expect the caller to have already verified this.  Would it be better to call this function dispatchNavigateEvent?

Where is it handled elsewhere?  Same question about download events.